tensorflow/core/common_runtime/zen_layout_pass.cc

```diff
@@ -0,0 +1,1292 @@
+    nb->Attr("explicit_paddings", explicit_paddings);  // sets padding values
+  }
+  // Padding attribute for condition when fusion is not performed
+  else {
+    nb->Attr("padding", padding);
+    // If padding is EXPLICIT, then explicit_paddings attribute needs to be set
+    if (padding == "EXPLICIT") {
+      std::vector<int32> explicit_paddings_tmp = {};
+      TF_CHECK_OK(GetNodeAttr(orig_node->def(), "explicit_paddings",
+                              &explicit_paddings_tmp));
+      nb->Attr("explicit_paddings", explicit_paddings_tmp);
+    }
+  }
+  nb->Attr("data_format", data_format);
+  nb->Attr("dilations", dilations);
+  nb->Attr("epsilon", epsilon);
+}
+
+static void FillInputs(const Node *n,
+                       gtl::InlinedVector<Node *, 4> *control_edges,
+                       gtl::InlinedVector<std::pair<Node *, int>, 4> *in) {
+  control_edges->clear();
+  for (const Edge *e : n->in_edges()) {
+    if (e->IsControlEdge()) {
+      control_edges->push_back(e->src());
+    } else {
+      (*in)[e->dst_input()] = std::make_pair(e->src(), e->src_output());
+    }
+  }
+  std::sort(control_edges->begin(), control_edges->end());
+}
+
+void ZenLayoutRewritePass::GetNodesProducingTFTensorList(
+    const gtl::InlinedVector<std::pair<Node *, int>, 4> &inputs, int *input_idx,
+    int list_length, std::vector<NodeBuilder::NodeOut> *output_nodes) {
+  CHECK_LT(*input_idx, inputs.size());
+  CHECK_GT(list_length, 0);
+  CHECK_NOTNULL(output_nodes);
+  output_nodes->reserve(list_length);
+
+  while (list_length != 0) {
+    CHECK_GT(list_length, 0);
+    CHECK_LT(*input_idx, inputs.size());
+    Node *n = inputs[*input_idx].first;
+    int slot = inputs[*input_idx].second;
+    // If input node 'n' is just producing a single tensor at
+    // output slot 'slot' then we just add that single node.
+    output_nodes->push_back(NodeBuilder::NodeOut(n, slot));
+    (*input_idx)++;
+    list_length--;
+  }
+}
+
+Status ZenLayoutRewritePass::ZenOpNodeRewrite(
+    std::unique_ptr<Graph> *g, Node *orig_node,
+    const ZenOpRewriteRecord *rewrite_record,
+    std::pair<bool, bool> reorder_flags) {
+  DCHECK_NE(rewrite_record, nullptr);
+  DCHECK_NE(orig_node, nullptr);
+
+  Status ret_status = OkStatus();
+  Node *new_node = nullptr;
+  std::vector<string> fused_ops = {};
+  int num_data_inputs = orig_node->in_edges().size();
+  for (const Edge *e : orig_node->in_edges()) {
+    if (e->IsControlEdge()) {
+      num_data_inputs--;
+    }
+  }
+
+  gtl::InlinedVector<Node *, 4> control_edges;
+  gtl::InlinedVector<std::pair<Node *, int>, 4> inputs(num_data_inputs);
+  FillInputs(orig_node, &control_edges, &inputs);
+
+  NodeBuilder nb(orig_node->name().c_str(),
+                 rewrite_record->zen_op_name.c_str());
+
+  nb.Device(orig_node->def().device());
+  ret_status = CopyInputs(orig_node, inputs, &nb);
+  if (ret_status != OkStatus()) {
+    return ret_status;
+  }
```

Comment:
We can use `TF_RETURN_IF_ERROR` for this. Same for other places.
```suggestion
  TF_RETURN_IF_ERROR(CopyInputs(orig_node, inputs, &nb));
```
(May need to include `tensorflow/core/platform/errors.h`.)

Reply:
Accepted change. Thank you!